doxygen/man/man3/MessageModel.3
.TH "MessageModel" 3 "Wed Feb 10 2016" "Version 1.0.0.0" "darksilk" \" -*- nroff -*-
.ad l
.nh
.SH NAME
MessageModel \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <messagemodel\&.h>\fP
.PP
Inherits QAbstractTableModel\&.
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBStatusCode\fP { \fBOK\fP, \fBInvalidAddress\fP, \fBInvalidMessage\fP, \fBDuplicateAddress\fP, \fBMessageCreationFailed\fP, \fBMessageCommitFailed\fP, \fBAborted\fP, \fBFailedErrorShown\fP }"
.br
.ti -1c
.RI "enum \fBColumnIndex\fP { \fBType\fP = 0, \fBSentDateTime\fP = 1, \fBReceivedDateTime\fP = 2, \fBLabel\fP = 3, \fBToAddress\fP = 4, \fBFromAddress\fP = 5, \fBMessage\fP = 6, \fBTypeInt\fP = 7, \fBKey\fP = 8, \fBHTML\fP = 9 }"
.br
.ti -1c
.RI "enum \fBRoleIndex\fP { \fBTypeRole\fP = Qt::UserRole, \fBKeyRole\fP, \fBSentDateRole\fP, \fBReceivedDateRole\fP, \fBFromAddressRole\fP, \fBToAddressRole\fP, \fBFilterAddressRole\fP, \fBLabelRole\fP, \fBMessageRole\fP, \fBShortMessageRole\fP, \fBHTMLRole\fP, \fBAmbiguous\fP }"
.br
.in -1c
.SS "Public Slots"

.in +1c
.ti -1c
.RI "void \fBnewMessage\fP (const \fBSecMsgStored\fP &smsg)"
.br
.ti -1c
.RI "void \fBnewOutboxMessage\fP (const \fBSecMsgStored\fP &smsg)"
.br
.ti -1c
.RI "void \fBwalletUnlocked\fP ()"
.br
.ti -1c
.RI "void \fBsetEncryptionStatus\fP (int status)"
.br
.in -1c
.SS "Signals"

.in +1c
.ti -1c
.RI "void \fBerror\fP (const QString &title, const QString &message, bool modal)"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBMessageModel\fP (\fBCWallet\fP *wallet, \fBWalletModel\fP *walletModel, QObject *parent=0)"
.br
.ti -1c
.RI "\fB~MessageModel\fP ()"
.br
.ti -1c
.RI "int \fBlookupMessage\fP (const QString &message) const "
.br
.ti -1c
.RI "\fBWalletModel\fP * \fBgetWalletModel\fP ()"
.br
.ti -1c
.RI "\fBOptionsModel\fP * \fBgetOptionsModel\fP ()"
.br
.ti -1c
.RI "void \fBresetFilter\fP ()"
.br
.ti -1c
.RI "bool \fBgetAddressOrPubkey\fP (QString &Address, QString &Pubkey) const "
.br
.ti -1c
.RI "\fBStatusCode\fP \fBsendMessages\fP (const QList< \fBSendMessagesRecipient\fP > &recipients)"
.br
.ti -1c
.RI "\fBStatusCode\fP \fBsendMessages\fP (const QList< \fBSendMessagesRecipient\fP > &recipients, const QString &addressFrom)"
.br
.in -1c
.PP
.RI "\fBMethods overridden from QAbstractTableModel\fP"
.br

.in +1c
.in +1c
.ti -1c
.RI "int \fBrowCount\fP (const QModelIndex &parent) const "
.br
.ti -1c
.RI "int \fBcolumnCount\fP (const QModelIndex &parent) const "
.br
.ti -1c
.RI "QVariant \fBdata\fP (const QModelIndex &\fBindex\fP, int role) const "
.br
.ti -1c
.RI "QVariant \fBheaderData\fP (int section, Qt::Orientation orientation, int role) const "
.br
.ti -1c
.RI "QModelIndex \fBindex\fP (int row, int column, const QModelIndex &parent) const "
.br
.ti -1c
.RI "bool \fBremoveRows\fP (int row, int count, const QModelIndex &parent=QModelIndex())"
.br
.ti -1c
.RI "Qt::ItemFlags \fBflags\fP (const QModelIndex &\fBindex\fP) const "
.br
.in -1c
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "QSortFilterProxyModel * \fBproxyModel\fP"
.br
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static const QString \fBSent\fP = 'Sent'"
.br
.ti -1c
.RI "static const QString \fBReceived\fP = 'Received'"
.br
.in -1c
.SS "Friends"

.in +1c
.ti -1c
.RI "class \fBMessageTablePriv\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Interface to DarkSilk Secure Messaging from Qt view code\&. 
.PP
Definition at line 71 of file messagemodel\&.h\&.
.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBMessageModel::ColumnIndex\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIType \fP\fP
Sent/Received 
.TP
\fB\fISentDateTime \fP\fP
Time Sent 
.TP
\fB\fIReceivedDateTime \fP\fP
Time Received 
.TP
\fB\fILabel \fP\fP
User specified label 
.TP
\fB\fIToAddress \fP\fP
To DarkSilk address 
.TP
\fB\fIFromAddress \fP\fP
From DarkSilk address 
.TP
\fB\fIMessage \fP\fP
Plaintext 
.TP
\fB\fITypeInt \fP\fP
Plaintext 
.TP
\fB\fIKey \fP\fP
chKey 
.TP
\fB\fIHTML \fP\fP
HTML Formatted Data 
.PP
Definition at line 91 of file messagemodel\&.h\&.
.SS "enum \fBMessageModel::RoleIndex\fP"
Roles to get specific information from a message row\&. These are independent of column\&. 
.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fITypeRole \fP\fP
Type of message 
.TP
\fB\fIKeyRole \fP\fP
Date and time this message was sent message key 
.TP
\fB\fISentDateRole \fP\fP
.TP
\fB\fIReceivedDateRole \fP\fP
Date and time this message was received 
.TP
\fB\fIFromAddressRole \fP\fP
From Address of message 
.TP
\fB\fIToAddressRole \fP\fP
To Address of message 
.TP
\fB\fIFilterAddressRole \fP\fP
Filter address related to message 
.TP
\fB\fILabelRole \fP\fP
Label of address related to message 
.TP
\fB\fIMessageRole \fP\fP
Full Message 
.TP
\fB\fIShortMessageRole \fP\fP
Short Message 
.TP
\fB\fIHTMLRole \fP\fP
HTML Formatted 
.TP
\fB\fIAmbiguous \fP\fP
Ambiguous bool 
.PP
Definition at line 107 of file messagemodel\&.h\&.
.SS "enum \fBMessageModel::StatusCode\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIOK \fP\fP
.TP
\fB\fIInvalidAddress \fP\fP
.TP
\fB\fIInvalidMessage \fP\fP
.TP
\fB\fIDuplicateAddress \fP\fP
.TP
\fB\fIMessageCreationFailed \fP\fP
.TP
\fB\fIMessageCommitFailed \fP\fP
.TP
\fB\fIAborted \fP\fP
.TP
\fB\fIFailedErrorShown \fP\fP
.PP
Definition at line 79 of file messagemodel\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "MessageModel::MessageModel (\fBCWallet\fP * wallet, \fBWalletModel\fP * walletModel, QObject * parent = \fC0\fP)\fC [explicit]\fP"

.PP
Definition at line 307 of file messagemodel\&.cpp\&.
.SS "MessageModel::~MessageModel ()"

.PP
Definition at line 322 of file messagemodel\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "int MessageModel::columnCount (const QModelIndex & parent) const"

.PP
Definition at line 433 of file messagemodel\&.cpp\&.
.SS "QVariant MessageModel::data (const QModelIndex & index, int role) const"

.PP
Definition at line 439 of file messagemodel\&.cpp\&.
.SS "void MessageModel::error (const QString & title, const QString & message, bool modal)\fC [signal]\fP"

.SS "Qt::ItemFlags MessageModel::flags (const QModelIndex & index) const"

.PP
Definition at line 512 of file messagemodel\&.cpp\&.
.SS "bool MessageModel::getAddressOrPubkey (QString & Address, QString & Pubkey) const"

.PP
Definition at line 331 of file messagemodel\&.cpp\&.
.SS "\fBOptionsModel\fP * MessageModel::getOptionsModel ()"

.PP
Definition at line 359 of file messagemodel\&.cpp\&.
.SS "\fBWalletModel\fP * MessageModel::getWalletModel ()"

.PP
Definition at line 354 of file messagemodel\&.cpp\&.
.SS "QVariant MessageModel::headerData (int section, Qt::Orientation orientation, int role) const"

.PP
Definition at line 507 of file messagemodel\&.cpp\&.
.SS "QModelIndex MessageModel::index (int row, int column, const QModelIndex & parent) const"

.PP
Definition at line 520 of file messagemodel\&.cpp\&.
.SS "int MessageModel::lookupMessage (const QString & message) const"

.SS "void MessageModel::newMessage (const \fBSecMsgStored\fP & smsg)\fC [slot]\fP"

.PP
Definition at line 558 of file messagemodel\&.cpp\&.
.SS "void MessageModel::newOutboxMessage (const \fBSecMsgStored\fP & smsg)\fC [slot]\fP"

.PP
Definition at line 564 of file messagemodel\&.cpp\&.
.SS "bool MessageModel::removeRows (int row, int count, const QModelIndex & parent = \fCQModelIndex()\fP)"

.PP
Definition at line 527 of file messagemodel\&.cpp\&.
.SS "void MessageModel::resetFilter ()"

.PP
Definition at line 553 of file messagemodel\&.cpp\&.
.SS "int MessageModel::rowCount (const QModelIndex & parent) const"

.PP
Definition at line 427 of file messagemodel\&.cpp\&.
.SS "\fBMessageModel::StatusCode\fP MessageModel::sendMessages (const QList< \fBSendMessagesRecipient\fP > & recipients)"

.PP
Definition at line 422 of file messagemodel\&.cpp\&.

CHANGELOG
Base58 Change, Port Changes and Minor Fixes
Addition of OBJ directory for Daemon builds
Version updated
ReadMe updated
Instamine code for Silk/Dark weaver swap/burn added to main.cpp
OpenSSL update
LevelDB 1.18 Update
Fix to guard against OpenSSL's new strict DER checks, fixed getbalance and other minor fixes
Spelling edits
Enabled locking of Stormnodes via the Stormnode configuration file
Unset Sandstorm checkbox if there were non-anonymized inputs selected in coincontrol, give warning
Removal of target space change and hardfork
Fixed time to switch "out of sync" messages
Fix twice denominatedBalance calculations
Solve performance problems by caching and fix probable security issue
RPC: Allow "sandstorm" only with ENABLE_WALLET
Disconnect Sandstorm timer on Overview page destruction
Fix spaces/monospace fonts in qt console
Mac compatibility so users can switch tabs using CMD+(NUM)
Added better instructions to get libsecp256k1 installed in case of library issues within ReadMe
Added secp256k1 libs to /deps in case for any reason the libs are not available from github anymore
Increased maximum block size to 50MB (50x Bitcoin Core)
Added safe shutdown for Windows
Fixed sync issues and CPU usage
Add event support for Tor
rpc: Prevent easy memory exhaustion attack 
Fixed issue where using setaccount with a foreign address causes the address to be added to your receiving addresses.
Fix priority calculation in CreateTransaction
darksilk.conf added to /src
GUI Styling changes
Statistics Page Added and Icons Updated
Clickable Padlock to Unlock Wallet and Unlock for Staking Only
Fixed GUI bugs
Assertion Error Fix for MacOSX
Sandstorm GUI Fixes
Decentralised DarkSilkMarket added
Overview Transactions UI Fixes
Messaging, GUI Update, Null Pointer Fix, Various Fixes/Changes
Windows Build Fix
Addition of possibly needed dependency
Stormnode Collateral Increase
Marketplace RPC Calls | Cleanup
Amended Spork Max Value for Stormnode Collateral
Removal of unused Spork
Minimum Transaction fee changed to 0.0001DRKSLK
Add support for miniupnpc API version 14
DARKSILK_TIMEDATA_MAX_SAMPLES 200
Add time offset to getpeerinfo output
Don't check duplicate stake while importing
Ignore blocks received while importing
Further Market RPC JSON calls
marketbuyrequest RPC command
marketmybuys RPC call for market
MacOSX Fixes and GUI Fix
Market RPC Minor Fixes
darksilk-qt.pro cleanup
DarkSilk-qt.app renamed to DarkSilk.app for OSX
Replaced stake min age with min confirmation
Limited drop of difficulty per block
Made special marker non spendable
Added nonspendable encoding for block signing key
BIP66 compatibility
Replaced NOP7 with CHECKLOCKTIMEVERIFY
Enforce new script verification flags
Static pos reward
Demand canonical block sig from updated peers
Submitblock: Add submission of stake template
Reward and Min Conf Blocks Changed
Checkpointing Removal
Gitian MacDeploy Graphics Added
IsProtocolV1/IsProtocolV2 hard fork switches removed
FutureDriftV1/FutureDriftV2 reverted back to FutureDrift
CheckStakeKernelHashV1/CheckStakeKernalHashV2 reverted back to CheckStakeKernelHash
Unsigned Int GetTargetSpacing changed to Static Const Int POS_TARGET_SPACING
Fix boost::get usage with Boost 1.58 in rpcrawtransaction.cpp 
Static PoS Reward of 1 DRKSLK
Prevent redefinition compiler warning
Fix issue where getrawtransaction RPC call does not work
PoW additions to GetNextTargetRequired
Scale up addrman
Gitian windows compile fixes
Correctly randomize change output position
LevelDB windows cross compile missing files
Scale up addrman
Gitian windows compile fixes
Correctly randomize change output position
Removal of Dead Code
Fix for building with Boost 1.58
LevelDB Cross compile fixes
DarkSilk-Qt: cleanup / optimise addressbookpage
Discourage fee sniping with nLockTime
Removal of dead code from main.h
Changed ProofOfStakeReward to use static const
Diff changes
Shadowchat Messaging Updates
DNSSeed Thread Fix

src/qt/messagemodel.cpp

#include "messagemodel.h"
#include "guiutil.h"
#include "guiconstants.h"
#include "darksilkunits.h"
#include "optionsmodel.h"
#include "walletmodel.h"
#include "addresstablemodel.h"
#include "ui_interface.h"
#include "base58.h"
#include "json_spirit.h"

#include <QSet>
#include <QTimer>
#include <QDateTime>
#include <QSortFilterProxyModel>
#include <QClipboard>
#include <QMessageBox>
#include <QMenu>
#include <QFont>
#include <QColor>

Q_DECLARE_METATYPE(std::vector<unsigned char>);

QList<QString> ambiguous; /**< Specifies Ambiguous addresses */

const QString MessageModel::Sent = "Sent";
const QString MessageModel::Received = "Received";

struct MessageTableEntryLessThan
{
    bool operator()(const MessageTableEntry &a, const MessageTableEntry &b) const {return a.received_datetime < b.received_datetime;};
    bool operator()(const MessageTableEntry &a, const QDateTime         &b) const {return a.received_datetime < b;}
    bool operator()(const QDateTime         &a, const MessageTableEntry &b) const {return a < b.received_datetime;}
};

// Private implementation
class MessageTablePriv
{
public:
    QList<MessageTableEntry> cachedMessageTable;
    MessageModel *parent;

    MessageTablePriv(MessageModel *parent):
        parent(parent) {}

    void refreshMessageTable()
    {
        cachedMessageTable.clear();
        
        if (parent->getWalletModel()->getEncryptionStatus() == WalletModel::Locked)
        {
            // -- messages are stored encrypted, can't load them without the private keys
            return;
        };

        {
            LOCK(cs_smsgDB);

            SecMsgDB dbSmsg;

            if (!dbSmsg.Open("cr+"))
                //throw runtime_error("Could not open DB.");
                return;

            unsigned char chKey[18];
            std::vector<unsigned char> vchKey;
            vchKey.resize(18);

            SecMsgStored smsgStored;
            MessageData msg;
            QString label;
            QDateTime sent_datetime;
            QDateTime received_datetime;

            std::string sPrefix("im");
            leveldb::Iterator* it = dbSmsg.pdb->NewIterator(leveldb::ReadOptions());
            while (dbSmsg.NextSmesg(it, sPrefix, chKey, smsgStored))
            {
                uint32_t nPayload = smsgStored.vchMessage.size() - SMSG_HDR_LEN;
                if (SecureMsgDecrypt(false, smsgStored.sAddrTo, &smsgStored.vchMessage[0], &smsgStored.vchMessage[SMSG_HDR_LEN], nPayload, msg) == 0)
                {
                    label = parent->getWalletModel()->getAddressTableModel()->labelForAddress(QString::fromStdString(msg.sFromAddress));

                    sent_datetime    .setTime_t(msg.timestamp);
                    received_datetime.setTime_t(smsgStored.timeReceived);
                    
                    memcpy(&vchKey[0], chKey, 18);

                    addMessageEntry(MessageTableEntry(vchKey,
                                                      MessageTableEntry::Received,
                                                      label,
                                                      QString::fromStdString(smsgStored.sAddrTo),
                                                      QString::fromStdString(msg.sFromAddress),
                                                      sent_datetime,
                                                      received_datetime,
                                                      (char*)&msg.vchMessage[0]),
                                    true);
                }
            };

            delete it;

            sPrefix = "sm";
            it = dbSmsg.pdb->NewIterator(leveldb::ReadOptions());
            while (dbSmsg.NextSmesg(it, sPrefix, chKey, smsgStored))
            {
                uint32_t nPayload = smsgStored.vchMessage.size() - SMSG_HDR_LEN;
                if (SecureMsgDecrypt(false, smsgStored.sAddrOutbox, &smsgStored.vchMessage[0], &smsgStored.vchMessage[SMSG_HDR_LEN], nPayload, msg) == 0)
                {
                    label = parent->getWalletModel()->getAddressTableModel()->labelForAddress(QString::fromStdString(smsgStored.sAddrTo));

                    sent_datetime    .setTime_t(msg.timestamp);
                    received_datetime.setTime_t(smsgStored.timeReceived);
                    
                    memcpy(&vchKey[0], chKey, 18);

                    addMessageEntry(MessageTableEntry(vchKey,
                                                      MessageTableEntry::Sent,
                                                      label,
                                                      QString::fromStdString(smsgStored.sAddrTo),
                                                      QString::fromStdString(msg.sFromAddress),
                                                      sent_datetime,
                                                      received_datetime,
                                                      (char*)&msg.vchMessage[0]),
                                    true);
                }
            };

            delete it;
        }
    }

    void newMessage(const SecMsgStored& inboxHdr)
    {
        // we have to copy it, because it doesn't like constants going into Decrypt
        SecMsgStored smsgStored = inboxHdr;
        MessageData msg;
        QString label;
        QDateTime sent_datetime;
        QDateTime received_datetime;

        uint32_t nPayload = smsgStored.vchMessage.size() - SMSG_HDR_LEN;
        if (SecureMsgDecrypt(false, smsgStored.sAddrTo, &smsgStored.vchMessage[0], &smsgStored.vchMessage[SMSG_HDR_LEN], nPayload, msg) == 0)
        {
            label = parent->getWalletModel()->getAddressTableModel()->labelForAddress(QString::fromStdString(msg.sFromAddress));

            sent_datetime    .setTime_t(msg.timestamp);
            received_datetime.setTime_t(smsgStored.timeReceived);

            std::string sPrefix("im");
            SecureMessage* psmsg = (SecureMessage*) &smsgStored.vchMessage[0];
            
            std::vector<unsigned char> vchKey;
            vchKey.resize(18);
            memcpy(&vchKey[0],  sPrefix.data(),  2);
            memcpy(&vchKey[2],  &psmsg->timestamp, 8);
            memcpy(&vchKey[10], &smsgStored.vchMessage[SMSG_HDR_LEN], 8);    // sample

            addMessageEntry(MessageTableEntry(vchKey,
                                              MessageTableEntry::Received,
                                              label,
                                              QString::fromStdString(smsgStored.sAddrTo),
                                              QString::fromStdString(msg.sFromAddress),
                                              sent_datetime,
                                              received_datetime,
                                              (char*)&msg.vchMessage[0]),
                            false);
        }
    }

    void newOutboxMessage(const SecMsgStored& outboxHdr)
    {

        SecMsgStored smsgStored = outboxHdr;
        MessageData msg;
        QString label;
        QDateTime sent_datetime;
        QDateTime received_datetime;

        uint32_t nPayload = smsgStored.vchMessage.size() - SMSG_HDR_LEN;
        if (SecureMsgDecrypt(false, smsgStored.sAddrOutbox, &smsgStored.vchMessage[0], &smsgStored.vchMessage[SMSG_HDR_LEN], nPayload, msg) == 0)
        {
            label = parent->getWalletModel()->getAddressTableModel()->labelForAddress(QString::fromStdString(smsgStored.sAddrTo));

            sent_datetime    .setTime_t(msg.timestamp);
            received_datetime.setTime_t(smsgStored.timeReceived);

            std::string sPrefix("sm");
            SecureMessage* psmsg = (SecureMessage*) &smsgStored.vchMessage[0];
            std::vector<unsigned char> vchKey;
            vchKey.resize(18);
            memcpy(&vchKey[0],  sPrefix.data(),  2);
            memcpy(&vchKey[2],  &psmsg->timestamp, 8);
            memcpy(&vchKey[10], &smsgStored.vchMessage[SMSG_HDR_LEN], 8);    // sample

            addMessageEntry(MessageTableEntry(vchKey,
                                              MessageTableEntry::Sent,
                                              label,
                                              QString::fromStdString(smsgStored.sAddrTo),
                                              QString::fromStdString(msg.sFromAddress),
                                              sent_datetime,
                                              received_datetime,
                                              (char*)&msg.vchMessage[0]),
                            false);
        }
    }

    void walletUnlocked()
    {
        // -- wallet is unlocked, can get at the private keys now
        refreshMessageTable();
        
        parent->reset(); // reload table view
        
        if (parent->proxyModel)
        {
            parent->proxyModel->setFilterRole(false);
            parent->proxyModel->setFilterFixedString("");
            parent->resetFilter();
            parent->proxyModel->setFilterRole(MessageModel::Ambiguous);
            parent->proxyModel->setFilterFixedString("true");
        }
        
        //invalidateFilter()
    }
    
    void setEncryptionStatus(int status)
    {
        if (status == WalletModel::Locked)
        {
            // -- Wallet is locked, clear secure message display.
            cachedMessageTable.clear();

            parent->reset(); // reload table view
        };
    };

    MessageTableEntry *index(int idx)
    {
        if(idx >= 0 && idx < cachedMessageTable.size())
            return &cachedMessageTable[idx];
        else
            return 0;
    }

private:
    // Get the json value
    const json_spirit::mValue & find_value(json_spirit::mObject & obj, const char * key)
    {
        std::string newKey = key;

        json_spirit::mObject::const_iterator i = obj.find(newKey);

        if(i != obj.end() && i->first == newKey)
            return i->second;
        else
            return json_spirit::mValue::null;
    }

    const std::string get_value(json_spirit::mObject & obj, const char * key)
    {
        json_spirit::mValue val = find_value(obj, key);

        if(val.is_null())
            return "";
        else
            return val.get_str();
    }

    // Determine if it is a special message, i.e.: Invoice, Receipt, etc...
    void handleMessageEntry(const MessageTableEntry & message, const bool append)
    {
        addMessageEntry(message, append);
        json_spirit::mValue mVal;
        json_spirit::read(message.message.toStdString(), mVal);

        if(mVal.is_null())
        {
            addMessageEntry(message, append);
            return;
        }

        json_spirit::mObject mObj(mVal.get_obj());
        json_spirit::mValue mvType = find_value(mObj, "type");

    }

    void addMessageEntry(const MessageTableEntry & message, const bool & append)
    {
        if(append)
        {
            cachedMessageTable.append(message);
        } else
        {
            int index = qLowerBound(cachedMessageTable.begin(), cachedMessageTable.end(), message.received_datetime, MessageTableEntryLessThan()) - cachedMessageTable.begin();
            parent->beginInsertRows(QModelIndex(), index, index);
            cachedMessageTable.insert(
                        index,
                        message);
            parent->endInsertRows();
        }
    }

};

MessageModel::MessageModel(CWallet *wallet, WalletModel *walletModel, QObject *parent) :
    QAbstractTableModel(parent), wallet(wallet), walletModel(walletModel), optionsModel(0), priv(0)
{
    columns << tr("Type") << tr("Sent Date Time") << tr("Received Date Time") << tr("Label") << tr("To Address") << tr("From Address") << tr("Message");
    
    proxyModel = NULL;
    
    optionsModel = walletModel->getOptionsModel();

    priv = new MessageTablePriv(this);
    priv->refreshMessageTable();

    subscribeToCoreSignals();
}

MessageModel::~MessageModel()
{
    if (proxyModel)
        delete proxyModel;
    
    delete priv;
    unsubscribeFromCoreSignals();
}

bool MessageModel::getAddressOrPubkey(QString &address, QString &pubkey) const
{
    CDarkSilkAddress addressParsed(address.toStdString());

    if(addressParsed.IsValid()) {
        CKeyID  destinationAddress;
        CPubKey destinationKey;

        addressParsed.GetKeyID(destinationAddress);

        if (SecureMsgGetStoredKey(destinationAddress, destinationKey) != 0
            && SecureMsgGetLocalKey(destinationAddress, destinationKey) != 0) // test if it's a local key
            return false;

        address = destinationAddress.ToString().c_str();
        pubkey = EncodeBase58(destinationKey.Raw()).c_str();

        return true;
    }

    return false;
}

WalletModel *MessageModel::getWalletModel()
{
    return walletModel;
}

OptionsModel *MessageModel::getOptionsModel()
{
    return optionsModel;
}

MessageModel::StatusCode MessageModel::sendMessages(const QList<SendMessagesRecipient> &recipients, const QString &addressFrom)
{

    QSet<QString> setAddress;

    if(recipients.empty())
        return OK;

    // Pre-check input data for validity
    foreach(const SendMessagesRecipient &rcp, recipients)
    {
        if(!walletModel->validateAddress(rcp.address))
            return InvalidAddress;

        if(rcp.message == "")
            return MessageCreationFailed;

        std::string sendTo  = rcp.address.toStdString();
        std::string pubkey  = rcp.pubkey.toStdString();
        std::string message = rcp.message.toStdString();
        std::string addFrom = addressFrom.toStdString();

        SecureMsgAddAddress(sendTo, pubkey);
        setAddress.insert(rcp.address);
        
        std::string sError;
        if (SecureMsgSend(addFrom, sendTo, message, sError) != 0)
        {
            QMessageBox::warning(NULL, tr("Send Secure Message"),
                tr("Send failed: %1.").arg(sError.c_str()),
                QMessageBox::Ok, QMessageBox::Ok);
            
            return FailedErrorShown;
        };

        // Add addresses / update labels that we've sent to to the address book
        std::string strAddress = rcp.address.toStdString();
        CTxDestination dest = CDarkSilkAddress(strAddress).Get();
        std::string strLabel = rcp.label.toStdString();
        {
            LOCK(wallet->cs_wallet);

            std::map<CTxDestination, std::string>::iterator mi = wallet->mapAddressBook.find(dest);

            // Check if we have a new address or an updated label
            if (mi == wallet->mapAddressBook.end() || mi->second != strLabel)
            {
                wallet->SetAddressBookName(dest, strLabel);
            }
        }
    }

    if(recipients.size() > setAddress.size())
        return DuplicateAddress;

    return OK;
}

MessageModel::StatusCode MessageModel::sendMessages(const QList<SendMessagesRecipient> &recipients)
{
    return sendMessages(recipients, "anon");
}

int MessageModel::rowCount(const QModelIndex &parent) const
{
    Q_UNUSED(parent);
    return priv->cachedMessageTable.size();
}

int MessageModel::columnCount(const QModelIndex &parent) const
{
    Q_UNUSED(parent);
    return columns.length();
}

QVariant MessageModel::data(const QModelIndex &index, int role) const
{
    if(!index.isValid())
        return QVariant();

    MessageTableEntry *rec = static_cast<MessageTableEntry*>(index.internalPointer());

    switch(role)
    {
    /*
    case Qt::DecorationRole:
        switch(index.column())
        {
            return txStatusDecoration(rec);
        case ToAddress:
            return txAddressDecoration(rec);
        }
        break;*/
    case Qt::DisplayRole:
        switch(index.column())
        {
            case Label:	           return (rec->label.isEmpty() ? tr("(no label)") : rec->label);
            case ToAddress:	       return rec->to_address;
            case FromAddress:      return rec->from_address;
            case SentDateTime:     return rec->sent_datetime;
            case ReceivedDateTime: return rec->received_datetime;
            case Message:          return rec->message;
            case TypeInt:          return rec->type;
            case HTML:             return rec->received_datetime.toString() + "<br>"  + (rec->label.isEmpty() ? rec->from_address : rec->label)  + "<br>" + rec->message;
            case Type:
                switch(rec->type)
                {
                    case MessageTableEntry::Sent:     return Sent;
                    case MessageTableEntry::Received: return Received;
                    default: break;
                }
            case Key:               return QVariant::fromValue(rec->chKey);
        }
        break;

    case KeyRole:           return QVariant::fromValue(rec->chKey);
    case TypeRole:          return rec->type;
    case SentDateRole:      return rec->sent_datetime;
    case ReceivedDateRole:  return rec->received_datetime;
    case FromAddressRole:   return rec->from_address;
    case ToAddressRole:     return rec->to_address;
    case FilterAddressRole: return (rec->type == MessageTableEntry::Sent ? rec->to_address + rec->from_address : rec->from_address + rec->to_address);
    case LabelRole:         return rec->label;
    case MessageRole:       return rec->message;
    case ShortMessageRole:  return rec->message; // TODO: Short message
    case HTMLRole:          return rec->received_datetime.toString() + "<br>"  + (rec->label.isEmpty() ? rec->from_address : rec->label)  + "<br>" + rec->message;
    case Ambiguous:
        int it;

        for (it = 0; it<ambiguous.length(); it++) {
            if(ambiguous[it] == (rec->type == MessageTableEntry::Sent ? rec->to_address + rec->from_address : rec->from_address + rec->to_address))
                return false;
        }
        QString address = (rec->type == MessageTableEntry::Sent ? rec->to_address + rec->from_address : rec->from_address + rec->to_address);
        ambiguous.append(address);

        return "true";
        break;
    }

    return QVariant();
}

QVariant MessageModel::headerData(int section, Qt::Orientation orientation, int role) const
{
    return (orientation == Qt::Horizontal && role == Qt::DisplayRole ? columns[section] : QVariant());
}

Qt::ItemFlags MessageModel::flags(const QModelIndex & index) const
{
    if(index.isValid())
        return Qt::ItemIsSelectable | Qt::ItemIsEnabled;

    return 0;
}

QModelIndex MessageModel::index(int row, int column, const QModelIndex & parent) const
{
    Q_UNUSED(parent);
    MessageTableEntry *data = priv->index(row);
    return (data ? createIndex(row, column, priv->index(row)) : QModelIndex());
}

bool MessageModel::removeRows(int row, int count, const QModelIndex & parent)
{
    MessageTableEntry *rec = priv->index(row);
    if(count != 1 || !rec)
        // Can only remove one row at a time, and cannot remove rows not in model.
        // Also refuse to remove receiving addresses.
        return false;

    {
        LOCK(cs_smsgDB);
        SecMsgDB dbSmsg;

        if (!dbSmsg.Open("cr+"))
            //throw runtime_error("Could not open DB.");
            return false;

        dbSmsg.EraseSmesg(&rec->chKey[0]);
    }

    beginRemoveRows(parent, row, row);
    priv->cachedMessageTable.removeAt(row);
    endRemoveRows();

    return true;
}

void MessageModel::resetFilter()
{
    ambiguous.clear();
}

void MessageModel::newMessage(const SecMsgStored &smsg)
{
    priv->newMessage(smsg);
}


void MessageModel::newOutboxMessage(const SecMsgStored &smsgOutbox)
{
    priv->newOutboxMessage(smsgOutbox);
}


void MessageModel::walletUnlocked()
{
    priv->walletUnlocked();
}

void MessageModel::setEncryptionStatus(int status)
{
    priv->setEncryptionStatus(status);
}


static void NotifySecMsgInbox(MessageModel *messageModel, SecMsgStored inboxHdr)
{
    // Too noisy: OutputDebugStringF("NotifySecMsgInboxChanged %s\n", message);
    QMetaObject::invokeMethod(messageModel, "newMessage", Qt::QueuedConnection,
                              Q_ARG(SecMsgStored, inboxHdr));
}

static void NotifySecMsgOutbox(MessageModel *messageModel, SecMsgStored outboxHdr)
{
    QMetaObject::invokeMethod(messageModel, "newOutboxMessage", Qt::QueuedConnection,
                              Q_ARG(SecMsgStored, outboxHdr));
}

static void NotifySecMsgWallet(MessageModel *messageModel)
{
    QMetaObject::invokeMethod(messageModel, "walletUnlocked", Qt::QueuedConnection);
}

void MessageModel::subscribeToCoreSignals()
{
    qRegisterMetaType<SecMsgStored>("SecMsgStored");

    // Connect signals
    NotifySecMsgInboxChanged.connect(boost::bind(NotifySecMsgInbox, this, _1));
    NotifySecMsgOutboxChanged.connect(boost::bind(NotifySecMsgOutbox, this, _1));
    NotifySecMsgWalletUnlocked.connect(boost::bind(NotifySecMsgWallet, this));
    
    connect(walletModel, SIGNAL(encryptionStatusChanged(int)), this, SLOT(setEncryptionStatus(int)));
}

void MessageModel::unsubscribeFromCoreSignals()
{
    // Disconnect signals
    NotifySecMsgInboxChanged.disconnect(boost::bind(NotifySecMsgInbox, this, _1));
    NotifySecMsgOutboxChanged.disconnect(boost::bind(NotifySecMsgOutbox, this, _1));
    NotifySecMsgWalletUnlocked.disconnect(boost::bind(NotifySecMsgWallet, this));
    
    disconnect(walletModel, SIGNAL(encryptionStatusChanged(int)), this, SLOT(setEncryptionStatus(int)));
}
